gufe/base.py

```diff
@@ -0,0 +1,108 @@
+import abc
+import uuid
+import sys
+import threading
+import hashlib
+import inspect
+from packaging.version import parse as parse_version
+from typing import Dict, Any, Callable
+
+
+class GufeTokenizableMixin:
+    """Mixin for all tokenizble tokenizeable gufe objects.
+
+    """
+
+    @property
+    def token(self):
+        if not hasattr(self, '_token') or self._token is None:
+            self._token = tokenize(self)
+        return self._token
+
+    @property
+    def key(self):
+        if not hasattr(self, '_key') or self._key is None:
+            prefix = type(self).__name__
+            self._key = GufeKey(f"{prefix}-{self.token}")
+        return self._key
+
+    @property
+    def defaults(self):
+        sig = inspect.signature(self.__init__)
+
+        defaults = {
+            param.name: param.default for param in sig.parameters.values()
+            if param.default is not inspect.Parameter.empty
+        }
+
+        return defaults
+
+    def _keyencode_dependencies(self, d):
+        for key, value in d.items():
+            if isinstance(value, dict):
+                self._keyencode_dependencies(value)
```

Comment:
this branch doesn't seem to catch the return value?

Reply:
Ah, good catch. Yeah this and the two that follow need to be reworked to look like `dictdecode_dependencies`. They're janky currently.

Reply:
Combined @dwhswenson's formulation and made encoders/decoders work recursively.